File: C-Plus-Plus/ds/PancakeSort.cpp
#include <iostream>
#include<bits/stdc++.h>
using namespace std;

//find out the maximum number in the array specified

int maximumNumber(int arr[], int curr_size)
{
  int max =0, index =0;
  
      for (int i = 0; i < curr_size; i++)
      {
        if (arr[i]>max)
        {
          max = arr[i];
          index = i;
        }
      }
  return index;
}

// Flip the array from start till the maximum index

int flip(int arr[], int max) 
{  
  int temp;
  
      for(int i = 0; i< max; i++)
      {
        temp = arr[i];
        arr[i] = arr[max];
        arr[max] = temp;
        max--;
      } 
}

//initiating the sorting algorithm

int PancakeSort(int arr[], int number)
{
  for (int curr_size = number; curr_size > 1 ; curr_size--)
  {
    int max = maximumNumber(arr, curr_size);
    
        if (max != curr_size -1)
        {
          flip(arr, max);               //flipping till the max element inclusive of max element
          flip(arr, curr_size-1);       // To place the max element at the end of the array and reduce the size by 1
        }
  }
  
  //Printing the sorted array
  
  cout<<"The sorted array is :  ";
      
      for (int i = 0; i < number; ++i)
          cout<<arr[i]<<" ";
     
  }

int main()
{
    int number;
    cin>>number;

    int arr[number];                          //creating an array
    for (int i = 0; i < number; ++i)         //input the array elements
        cin>>arr[i];
      
    PancakeSort(arr,number);
    return 0;
}

File: JavaScript/sort/PancakeSort.js
/*

Unlike a traditional sorting algorithm, which attempts to sort with the fewest comparisons possible,
the pancake sorting algorithm sorts the sequence in as few reversals as possible. Something similar to selection sort.

Algorithm - Start from current size equal to n and reduce current size by one while it's greater than 1.
                1. Find max form the reduced array size.
                2. Compare max with array elements and flip accordingly

Time Complexity - O(n ^ 2)    
            
*/

// Prompt node package for taking user input
const prompt = require("prompt-sync")({ sigint: true });

// Flip elements
function flip(array, index) {
  let start = 0;
  while (start < index) {
    [array[start], array[index]] = [array[index], array[start]];
    start++;
    index--;
  }
}

// Perform pancake sorting
function panCake(array) {

  // Create chunks of array one by one
  for (let i = array.length; i > 1; --i) {
    let max = array.indexOf(Math.max(...array.slice(0, i))); // get the max of chunk

    if (max !== i - 1) {
      flip(array, max); // flip array by moving max to the beginning
      flip(array, i - 1); // reverse the array
    }
  }

}

/* Workflow of user input */

// Take array length as input
let arrayLength = +prompt("Enter array length - ");

// Check whether the entered value is number or not
if (isNaN(arrayLength)) return console.log("Only numbers are allowed");

// Globally declared array
let array = [];

// Take array items
for (let i = 1; i <= arrayLength; i++) {
  array.push(+prompt(`Enter ${i} element - `));

  // Check whether the entered value is number or not
  if (array.includes(NaN)) return console.log("Only numbers are allowed");
}

console.log("Your array - ", array);

// Call the algorithm
panCake(array);
console.log("Sorted array - ", array);

/*

> node PancakeSort

Enter array length - 5
Enter 1 element - 21
Enter 2 element - 50
Enter 3 element - 11
Enter 4 element - 90
Enter 5 element - 45
Your array -  [ 21, 50, 11, 90, 45 ]
Sorted array - [ 11, 21, 45, 50, 90 ]

*/

